validator_client/src/block_service.rs

```diff
@@ -314,118 +323,245 @@ impl<T: SlotClock + 'static, E: EthSpec> BlockService<T, E> {
+                            BlockType::Full => {
+                                let _get_timer = metrics::start_timer_vec(
+                                    &metrics::BLOCK_SERVICE_TIMES,
+                                    &[metrics::BEACON_BLOCK_HTTP_GET],
+                                );
+                                beacon_node
+                                    .get_validator_blocks::<E, Payload>(
+                                        slot,
+                                        randao_reveal_ref,
+                                        graffiti.as_ref(),
+                                    )
+                                    .await
+                                    .map_err(|e| {
+                                        BlockError::Recoverable(format!(
+                                            "Error from beacon node when producing block: {:?}",
+                                            e
+                                        ))
+                                    })?
+                                    .data
+                            }
+                            BlockType::Blinded => {
+                                let _get_timer = metrics::start_timer_vec(
+                                    &metrics::BLOCK_SERVICE_TIMES,
+                                    &[metrics::BLINDED_BEACON_BLOCK_HTTP_GET],
+                                );
+                                beacon_node
+                                    .get_validator_blinded_blocks::<E, Payload>(
+                                        slot,
+                                        randao_reveal_ref,
+                                        graffiti.as_ref(),
+                                    )
+                                    .await
+                                    .map_err(|e| {
+                                        BlockError::Recoverable(format!(
+                                            "Error from beacon node when producing block: {:?}",
+                                            e
+                                        ))
+                                    })?
+                                    .data
+                            }
+                        };
+
+                        if proposer_index != Some(block.proposer_index()) {
+                            return Err(BlockError::Recoverable(
+                            "Proposer index does not match block proposer. Beacon chain re-orged"
+                                .to_string(),
+                        ));
+                        }
+
+                        Ok::<_, BlockError>(block)
+                    },
+                )
+                .await
+            {
+                Ok(v) => Ok(v), // Return the first block
+                Err(e) => {
+                    // There was an error try the proposer nodes.
+                    // TODO: Find a pattern to avoid this code duplication.
```

Comment:
I have a strategy in mind to avoid the code duplication and I'd like to give it a shot before it gets merged. I'll report back here with progress.

Reply:
I've created a PR this this branch over at: https://github.com/AgeManning/lighthouse/pull/4

That PR merges `unstable` and then removes the duplication. It reduces the lines added by this PR from 585 to 440.